File: Script/QuickSort/QuickSort.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <unistd.h>   // chdir, readlink
#include <libgen.h>   // dirname
#include <string.h>   // strerror
#include <errno.h>

// Funzione di utilità per lo scambio (swap)
void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

// Funzione di Partizione (Partition)
int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);

    for (int j = low; j <= high - 1; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}

// Funzione principale di Quick Sort
void quick_sort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}

// Funzione per la stampa dell'array
void printArray(int A[], int size) {
    for (int i = 0; i < size; i++)
        printf("%d ", A[i]);
    printf("\n");
}

// --- Gestione Input/Output ---
int main(void) {
    // Porta il working directory nella cartella dell’eseguibile
    char exe_path[PATH_MAX];
    ssize_t len = readlink("/proc/self/exe", exe_path, sizeof(exe_path) - 1);
    if (len == -1) {
        fprintf(stderr, "Errore: readlink fallita: %s\n", strerror(errno));
        return 1;
    }
    exe_path[len] = '\0';
    char *dir = dirname(exe_path);
    if (chdir(dir) != 0) {
        fprintf(stderr, "Errore: chdir(%s) fallito: %s\n", dir, strerror(errno));
        return 1;
    }

    FILE *file = fopen("input.txt", "r");
    if (!file) {
        fprintf(stderr, "Errore: impossibile aprire 'input.txt' in %s: %s\n", dir, strerror(errno));
        return 1;
    }

    int *arr = NULL;
    int count = 0;
    int number;

    while (fscanf(file, "%d", &number) == 1) {
        int *tmp = realloc(arr, (count + 1) * sizeof(int));
        if (!tmp) {
            fprintf(stderr, "Errore di allocazione della memoria\n");
            free(arr);
            fclose(file);
            return 1;
        }
        arr = tmp;
        arr[count++] = number;
    }
    fclose(file);

    if (count == 0) {
        fprintf(stderr, "Il file 'input.txt' è vuoto o non contiene numeri validi.\n");
        free(arr);
        return 0;
    }

    printf("Array non ordinato: ");
    printArray(arr, count);

    quick_sort(arr, 0, count - 1);

    printf("Array ordinato: ");
    printArray(arr, count);

    free(arr);
    return 0;
}

File: Script/QuickSort/QuickSort.rs
use std::fs;
use std::path::Path;

// Funzione principale di Quick Sort (lavora in-place su una slice mutabile)
fn quick_sort(arr: &mut [i32]) {
    if arr.len() > 1 {
        // Partiziona e ottieni l'indice del pivot
        let pivot_index = partition(arr);
        
        // Ordina la parte sinistra ricorsivamente
        quick_sort(&mut arr[0..pivot_index]);
        
        // Ordina la parte destra ricorsivamente
        quick_sort(&mut arr[pivot_index + 1..]);
    }
}

// Funzione di Partizione
fn partition(arr: &mut [i32]) -> usize {
    let len = arr.len();
    // Scegli l'elemento più a destra come pivot
    let pivot_value = arr[len - 1]; 
    let mut i = 0; // Puntatore per la sezione minore del pivot

    for j in 0..len - 1 {
        if arr[j] <= pivot_value {
            // Scambia l'elemento corrente con l'elemento in posizione i
            arr.swap(i, j); 
            i += 1;
        }
    }
    // Sposta il pivot nella sua posizione finale
    arr.swap(i, len - 1);
    i // Ritorna l'indice finale del pivot
}

// --- Gestione Input/Output ---
fn main() {
    let path = Path::new("input.txt");
    
    let content = match fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => {
            println!("Errore: il file 'input.txt' non è stato trovato.");
            return;
        }
    };

    // Parsa i dati
    let mut arr: Vec<i32> = content
        .split_whitespace()
        .filter_map(|s| s.parse::<i32>().ok())
        .collect();

    if arr.is_empty() {
        println!("Il file 'input.txt' è vuoto o non contiene numeri validi.");
        return;
    }

    println!("Array non ordinato: {:?}", arr);
    
    // Esegue il Quick Sort, passando un riferimento mutabile al vettore
    quick_sort(&mut arr);
    
    println!("Array ordinato: {:?}", arr);
}